File: exercism/python/robot-name/robot_name.py
#!/usr/bin/python3
# https://exercism.io/my/solutions/34f68f29449346e59df4e6d94d2f217f

from itertools import product
from random import shuffle
from string import ascii_uppercase, digits


letters = [''.join(p) for p in product(ascii_uppercase, ascii_uppercase)]
numbers = [''.join(p) for p in product(digits, digits, digits)]
names = [l + n for l, n in product(letters, numbers)]
shuffle(names)
NAMES = iter(names)


class Robot(object):
    def __init__(self):
        self.reset()

    def reset(self):
        try:
            self.name = next(NAMES)
        except StopIteration:
            raise Exception("No unique names available.")

File: exercism/go/robot-name/robot_name.go
// https://exercism.io/my/solutions/b176b173287d4e2598b135de4b92f0ec

package robotname

import (
	"errors"
	"fmt"
	"math/rand"
	"time"
)

const maxNum = 1000

var numPossibleNames = 26 * 26 * maxNum
var usedNames = map[string]bool{}
var random = rand.New(rand.NewSource(time.Now().UnixNano()))

// Robot represents a named robot
type Robot struct {
	name string
}

// Name returns a robot's name, assigning it one first if needed.
func (r *Robot) Name() (string, error) {
	if r.name == "" {
		if len(usedNames) >= numPossibleNames {
			return "", errors.New("no more unique names")
		}
		r.name = randomName()
		for usedNames[r.name] {
			r.name = randomName()
		}
		usedNames[r.name] = true
	}
	return r.name, nil
}

// Reset clears a robot's name
func (r *Robot) Reset() {
	r.name = ""
}

// randomName returns a pseudo-random string in the form [A-Z][A-Z][0-9][0-9][0-9]
func randomName() string {
	return fmt.Sprintf("%c%c%03d",
		rune(random.Intn(26)+int('A')),
		rune(random.Intn(26)+int('A')),
		random.Intn(maxNum),
	)
}
